---
title: "Exercise - Annotation Resources"
author: "Marc A.T. Teunis"
date: '`r Sys.Date()`'
output: 
   BiocStyle::html_document:
    toc: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, 
                      warning = FALSE, 
                      message = FALSE, 
                      error = FALSE,
                      results = 'hide',
                      fig.show = 'hide')
``` 

# Use relative paths
```{r, eval=TRUE, results='markup', echo=TRUE}
if (!require("rprojroot")) install.packages("rprojroot")
library("rprojroot")
root <- find_root_file(criterion = is_rstudio_project)
root
```

# End product

**Generate a new RMarkdown file containing the answers to all questions below and render it to HTML. Upload your _Rmd AND HTML_ file to CANVAS under the corresponding (see file name) exercise number** 

# Prerequisites

 - Go over the demo "Annotation"

# Packages
```{r}
suppressPackageStartupMessages({
    library(org.Hs.eg.db)
    library(TxDb.Hsapiens.UCSC.hg38.knownGene)
    library(EnsDb.Hsapiens.v75)
    library(BSgenome.Hsapiens.UCSC.hg38)
    library(GenomicRanges)
    library(biomaRt)
    library(rtracklayer)
    library(Gviz)
    library(AnnotationHub)
    library(tidyverse)
})


```

## Original Authors: Martin Morgan, Sonali Arora
Adaptations: Marc Teunis

## Aim
**Objective**: Learn about _Bioconductor_ resources for gene and
  genome annotation.

**Lessons learned**: 

- Use `org.*` packages for mapping between gene symbols.
- Use `TxDb.*` and `ensembldb` (`EnsDb.*`) packages for working with gene
  models.
- Use `AnnotationHub` to easily obtain select consortium-level resources
- Access `biomaRt` and other internet-based resources for highly
  flexible annotation.
- Use `VariantAnnotation` and `VariantFiltering` for annotating SNPs.

# Gene annotation

## Data packages

Organism-level ('org') packages contain mappings between a central
identifier (e.g., Entrez gene ids) and other identifiers (e.g. GenBank
or Uniprot accession number, RefSeq id, etc.).  The name of an org
package is always of the form `org.<Sp>.<id>.db`
(e.g. [org.Sc.sgd.db][]) where `<Sp>` is a 2-letter abbreviation of
the organism (e.g. `Sc` for *Saccharomyces cerevisiae*) and `<id>` is
an abbreviation (in lower-case) describing the type of central
identifier (e.g. `sgd` for gene identifiers assigned by the
*Saccharomyces* Genome Database, or `eg` for Entrez gene ids).  The
"How to use the '.db' annotation packages" vignette in the
[AnnotationDbi][] package (org packages are only one type of ".db"
annotation packages) is a key reference.  The '.db' and most other
Bioconductor annotation packages are updated every 6 months.

## `AnnotationDb` objects
Annotation packages usually contain an object named after the package
itself.  These objects are collectively called `AnnotationDb` objects,
with more specific classes named `OrgDb`, `ChipDb` or `TranscriptDb`
objects.  Methods that can be applied to these objects include
`cols()`, `keys()`, `keytypes()` and `select()`.  Common operations
for retrieving annotations are summarized in the table.

| Category   | Function                              | Description                                                      |
|------------|---------------------------------------|------------------------------------------------------------------|
| Discover   | `columns()`                           | List the kinds of columns that can be returned                   |
|            | `keytypes()`                          | List columns that can be used as keys                            |
|            | `keys()`                              | List values that can be expected for a given keytype             |
|            | `select()`                            | Retrieve annotations matching `keys`, `keytype` and `columns`    |
| Manipulate | `setdiff()`, `union()`, `intersect()` | Operations on sets                                               |
|            | `duplicated()`, `unique()`            | Mark or remove duplicates                                        |
|            | `%in%`,  `match()`                    | Find matches                                                     |
|            | `any()`, `all()`                      | Are any `TRUE`?  Are all?                                        |
|            | `merge()`                             | Combine two different \Robject{data.frames} based on shared keys |
| `GRanges*` | `transcripts()`, `exons()`, `cds()`   | Features (transcripts, exons, coding sequence) as `GRanges`.     |
|            | `transcriptsBy()` , `exonsBy()`       | Features group by  gene, transcript, etc., as `GRangesList`.     |
|            | `cdsBy()`                             |                                                                  |

## Internet resources

A short summary of select Bioconductor packages enabling web-based
queries is in following Table.

| Package                                             | Description                               |
|-----------------------------------------------------|-------------------------------------------|
| [AnnotationHub][]                                   | Ensembl, Encode, dbSNP, UCSC data objects |
| [biomaRt](http://biomart.org)                       | Ensembl and other annotations             |
| [PSICQUIC](https://code.google.com/p/psicquic)      | Protein interactions                      |
| [uniprot.ws](http://uniprot.org)                    | Protein annotations                       |
| [KEGGREST](http://www.genome.jp/kegg)               | KEGG pathways                             |
| [SRAdb](http://www.ncbi.nlm.nih.gov/sra)            | Sequencing experiments.                   |
| [rtracklayer](http://genome.ucsc.edu)               | genome tracks.                            |
| [GEOquery](http://www.ncbi.nlm.nih.gov/geo/)        | Array and other data                      |
| [ArrayExpress](http://www.ebi.ac.uk/arrayexpress/)  | Array and other data                      |

# Exercises

# **Exercise 1**: This exercise illustrates basic use of the `select'
interface to annotation packages.

## Question 1. Install and attach the [org.Hs.eg.db][] annotation package; it
   contains 'symbol mapping' information for _Homo sapiens_, based on
   NCBI 'Entrez' identifiers.

```{r}
library(org.Hs.eg.db)
```

## Question 2. Take a quick look at a summary of data in this package

```{r}
org.Hs.eg.db
```
    
### Question 3. The idea is that there are `keytypes()` that can be mapped to
   different `columns()`; `keys()` can be used to see available
   keys. Explore the package to see what sorts of information is
   available, e.g.,
  
Use the functions below:
  
  - `keytypes()`
  - `columns()`
  - `head()`
  
```{r}
keytypes(org.Hs.eg.db)
columns(org.Hs.eg.db)
head(keys(org.Hs.eg.db, "SYMBOL"))
```
   
## Question 4. There are two basic ways of extracting data from an `{org.*}` package
   -- `mapIds()` to create a 1:1 mapping between key and a single
   column, and `select()` (it's often necessary to specify this
   function directly, to avoid a conflict with dplyr, as
   `AnnotationDbi::select()`). Explore these functions, e.g.,
   
Run the code below to load 60 random example features ("ENTREZID") from the `org.Hs.eg.db` package (stored as a file in "data"). The commented code is how the file was created, you do not need to run the commented code.    

```{r, echo=TRUE}
# set.seed(123)

# egid <- sample(keys(org.Hs.eg.db), 60) %>%
#  sort() %>%
#  as_tibble() %>%
#  dplyr::rename(entrezgene = value) %>%
#  arrange(entrezgene)
# write_tsv(egid, path = file.path(root, "data", "entrezgenes.tsv"))

egid <- read_tsv(file = file.path(root, "data", "entrezgenes.tsv"), col_types = cols(entrezgene = col_character()))
egid
```  

## Question 5; Annotate

A) Use `mapIds()` to find all "SYMBOL" annotations for each "ENTREZID" in `egid`    

B) Use `AnnotationDbi::select()` together with `as_tibble()` get a dataframe in which "ENTREZID" is annotated with its equivalent "SYMBOL", its "GENENAME" and "GO" 

**TIPS**

 - Look at the help function for each function to be used to learn more
 - You have to use the `egid` object from above to get to the answer
 - Use the `%>%` in combination with `as_tibble()`
 - Store the table in a variable in your .GlobalEnv

```{r}    

## A)
mappings <- mapIds(
  org.Hs.eg.db, egid$entrezgene, 
          "ENSEMBL", 
          "ENTREZID")

mappings

annot_annotation_dbi <- AnnotationDbi::select(
        org.Hs.eg.db, egid$entrezgene, 
        c("ENSEMBL")) %>% 
          as_tibble()

annot_annotation_dbi


## B)
annot_annotation_dbi <- AnnotationDbi::select(
        org.Hs.eg.db, egid$entrezgene, 
        c("SYMBOL", 
          "ENSEMBL", 
          "GENENAME", 
          "ENTREZID",
          "GO")) %>% 
          as_tibble()

annot_annotation_dbi
```
   
 D) Some key - column mappings are 1:many, e.g., Entrez ID `"3123"` maps to a number of Ensembl Ids. What does `mapIds()` return when mapping Entrez ID `"3812"` to Ensembl ids? Use the additional argument `multiVals = "CharacterList"` to explore further. Compare results to those returned by `AnnotationDbi::select()`.

**TIPS**
 
 - Can you think of a way to use `dplyr::filter()` here? 
 - Try using `as_tibble()` to get a meaningful output
   
```{r}
egid_many <- "3123"

## maps the first ensembl_id
mapIds(org.Hs.eg.db, egid_many, "ENSEMBL", "ENTREZID")

## maps all ensembl_ids  
mappings_many <- mapIds(
  org.Hs.eg.db, 
  egid_many, 
  "ENSEMBL", 
  "ENTREZID",
  multiVals = "CharacterList"
)

mappings_many %>% as_tibble()
    
## or  
AnnotationDbi::select(
  org.Hs.eg.db, 
  egid_many, 
  c("SYMBOL", "ENSEMBL"),
  multiVals = "CharacterList"
)
    
## or
annot_annotation_dbi %>%
  dplyr::select(ENTREZID, ENSEMBL, SYMBOL) %>%
  dplyr::filter(ENTREZID == egid_many) %>%
  unique()
    
```

## Question 6; Tidyverse pipeline for Annotation

It seems like it might often be useful to use the tidyverse on return values from `mapIds()` and `AnnotionDbi::select()`; explore this usage further. Use the pipe (`%>%` ) to construct a valid pipeline. Use **ALL** the entrez ids available in the `{org.Hs.eg.db}` package. 

**TIPS**

 - The `keys(x = {org*}, keys = "ENTREZID")` function gives you all the available keys from a certain package 
 - Use attributes (stuff you get back) "SYMBOL", "ENTREZID", "ENSEMBL", "GO", "GENENAME"
 
   
```{r, message=FALSE}
library(tidyverse)
egid <- keys(x = org.Hs.eg.db, keys = "ENSEMBL")    # all ENTREZIDs

all <-  mapIds(
  org.Hs.eg.db, 
  egid, 
  "SYMBOL", 
  "ENTREZID") 

## converting an object from mapIds to a nice tibble is somewhat clumsy ... 
all %>%
  as.matrix() %>%
  as.data.frame() %>% 
  dplyr::mutate(ENTREZID = rownames(.),
                SYMBOL = V1) %>%
  dplyr::select(ENTREZID, SYMBOL) %>%
  head()
    
## better use AnnotationDbi if you are combining with tidyverse tibbles !!!       
AnnotationDbi::select(
  org.Hs.eg.db, egid, c("SYMBOL", "GO", "GENENAME"), "ENTREZID") %>% 
  as_tibble()
```

# **Exercise 2**: Using [biomaRt][].

## Question 1
 A) Explore the Biomart web site https://www.ensembl.org/biomart for retrieving all kinds of genomic annotations. 
 
Start by choosing a database (e.g., 'Ensembl Genes 92'), dataset (e.g., 'Human genes (GRCh38.p12)'), filter (e.g., 'GENE' / 'Input external reference' / 'Gene stable id' and enter
'ENSG00000000003'), attributes (default is ok), then press 'Results' to map from Ensembl identifier to transcript identifier.
   
 B) Load the `{biomaRt}` package. Use `listMarts()` to see availble databases, `useMart()` to select the mart you're interested in. Select 'ENSEMBL_MART_ENSEMBL'

```{r}
library(biomaRt)
head(listMarts())
mart <- useMart("ENSEMBL_MART_ENSEMBL")
```
   
 C) Use `listDatasets()` and `useDataset()` to select the _Homo sapiens_ gene dataset.
   
```{r}
head(listDatasets(mart))
dataset <- useDataset("hsapiens_gene_ensembl", mart)
```
   
 D) Use `listFilters()` to see available filters. The filter is the type of data that you are querying with (**stuff you put in**). Choose "ensembl_gene_id" and store it in a variable.

```{r}
head(listFilters(dataset))
filters <- "ensembl_gene_id"                    # see `listFilters()`
```

 E) Use `listAttrbutes()` to see available attributes. Attributes represent the information you'd like to retrieve (**stuff you get out**). Choose `c("ensembl_gene_id", "hgnc_symbol", "go_id", "description")` and create a variable to store these.
   
```{r}
head(listAttributes(dataset))
attrs <- c("ensembl_gene_id", "hgnc_symbol", "go_id", "description")    # see `listAttributes()`
```

 F) Create a character vector of 99 human Ensembl gene ids (from the list below), compose and execute the query (`biomaRt::getBM()`), transforming the result to a tibble. Use the ENSEMBL ids that map to ENTREZID and to SYMBOL `r egid_many` 
 
To get 99 ENSEMBL IDS:
```{r, echo=TRUE}
set.seed(123)
x <- AnnotationDbi::select(
  org.Hs.eg.db, egid, c("ENSEMBL"), "ENTREZID") %>% 
  as_tibble() %>%
  na.omit()
ensembl_99 <- sample_n(x, 99)$ENSEMBL
```

```{r}
## the Biomart quesry with getBM()
tbl <- getBM(attributes = attrs, 
             filters = filters, 
             values = ensembl_99, dataset) %>% as_tibble()
tbl
```

# **Exercise 3**: [KEGGREST][]

<font color="red">Internet access required for this exercise</font>

## Question 1; KEGGREST package exploration

 A) Explore the KEGG web site https://www.genome.jp/kegg/ KEGG is a database of information on pathways.
    
 B) Load the [KEGGREST][] package and discover available databases (`listDatabases()`)

```{r}
library(KEGGREST)
KEGGREST::listDatabases()
```
    
 C) Use `keggList()` to query the pathway database for human pathways ("pathway", "hsa"); present the result as a tibble. Run the following code

**TIPS**   
`hsa_pathways <- keggList("pathway", "hsa")`   
   
```{r, echo=TRUE}
hsa_pathways <- keggList("pathway", "hsa") %>% 
  tibble(pathway = names(.), description = .)
hsa_pathways
```
   
 D) Use `keggLink()` to recover the genes in each pathway. Use `group_by(pathway)` and `summarize(genes = list(egid))` to get the genes per pathway. 

```{r}
hsa_path_eg  <- keggLink("pathway", "hsa") %>% 
  tibble(pathway = ., egid = sub("hsa:", "", names(.)))
  hsa_path_eg
hsa_path_eg %>% group_by(pathway) %>% summarize(genes = list(egid))
```

 E) Update the `hsa_path_eg` table to include information on gene symbol and Ensembl id from the `org.Hs.eg.db` package. Retrieve the relevant information using `mapIds()`. How would you deal with entrez gene ids that map to multiple Ensembl ids?

```{r}
hsa_kegg_anno <- hsa_path_eg %>%
 dplyr::mutate(
   symbol = mapIds(org.Hs.eg.db, egid, "SYMBOL", "ENTREZID"),
   ensembl = mapIds(org.Hs.eg.db, egid, "ENSEMBL", "ENTREZID")
)
hsa_kegg_anno
```

 F) Use `left_join()` to append pathway descriptions to the
   `hsa_kegg_anno` table.

```{r}
x <- left_join(hsa_kegg_anno, hsa_pathways)

## remove duplicated
x[duplicated(x$egid),]
```

[KEGGREST]: https://bioconductor.org/packages/KEGGREST

# Genome annotation

There are a diversity of packages and classes available for
representing large genomes. Several include:

- `TxDb.*` and `EnsDb.*` For transcript and other genome / coordinate
  annotation.
- [BSgenome][] For whole-genome representation. See
  `available.genomes()` for pre-packaged genomes, and the vignette
  'How to forge a BSgenome data package' in the
- [Homo.sapiens][] For integrating 'TxDb*' and 'org.*' packages.
- 'SNPlocs.*' For model organism SNP locations derived from dbSNP.
- `FaFile()` ([Rsamtools][]) for accessing indexed FASTA files.
- [ensemblVEP][] Variant effect scores.

## Transcript annotation packages

Genome-centric packages are very useful for annotations involving
genomic coordinates. It is straight-forward, for instance, to discover
the coordinates of coding sequences in regions of interest, and from
these retrieve corresponding DNA or protein coding sequences. Other
examples of the types of operations that are easy to perform with
genome-centric annotations include defining regions of interest for
counting aligned reads in RNA-seq experiments and retrieving DNA
sequences underlying regions of interest in ChIP-seq analysis, e.g.,
for motif characterization.

## _rtracklayer_

The [rtracklayer][] package allows us to query the UCSC genome
browser, as well as providing `import()` and `export()` functions for
common annotation file formats like GFF, GTF, and BED. The exercise
below illustrates some of the functionality of [rtracklayer][].

# Exercises

# **Exercise 4**: `TxDb.*` packages

## Question 1. 

A) Load the [TxDb.Hsapiens.UCSC.hg38.knownGene][] package. This contains the gene models for _Homo sapiens_ based on the 'hg38' build of the human genome, using gene annotations in the UCSC 'knownGene' annotation track; TxDb's for more recent builds and for different annotation tracks are available. Take a look at a summary of the package, and create an alias for easy typing
   
```{r, echo=TRUE}
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
TxDb.Hsapiens.UCSC.hg38.knownGene
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene
```
      
B) The main purpose of this package is to provide genomic coordinates of genomic features such as `exons()`, coding sequences (`cds()`), `transcripts()` and `genes()`. Explore, for example,
```{r, echo=TRUE}
ex <- exons(txdb)
ex
library(ggplot2)
qplot(log10(width(ex)))
ex[ which.max(width(ex)) ]
```
   
C) Extract all genes, and then keep only the 'standard' chromosomes 1:22, X, Y,  Use `table()` and `seqnames()` to determine how many genes are on each chromosome. Also do this in a dplyr way; note that the `seqnames(gn)` need to be coerced with `as.factor()`. Create a plot to show the number of genes for each chomosome.

```{r, echo=TRUE}
gn <- genes(txdb)
gn_tbl <- as_tibble(gn)
levels(gn_tbl$seqnames)
length(gn)
std <- paste0("chr", c(1:22, "X", "Y"))
seqlevels(gn, pruning.mode = "coarse") <- std
length(gn)
seqlevels(gn)
table( seqnames(gn) )
data <- tibble(chr = as.factor(seqnames(gn))) %>% 
  group_by(chr) %>% 
  summarize(n = n())

```

**Code for plot -->**
Rank the chromosomes from fewest genes to most genes in the plot. Use `reorder()`
```{r}
data %>%
  ggplot(aes(x = reorder(as_factor(chr), n), y = n)) +
  geom_point() +
  theme_bw() +
   theme(axis.text.x = element_text(angle = -90, hjust = 1))
```

The following code plots the different chromosomes for the human:
```{r}
library(karyoploteR)
 kp <- plotKaryotype(genome="hg19", plot.type=1, main="The H. sapiens genome")
```

 D) The lenth of the chromsomes differ. The following code retrieves the approximate length of each chromosome (start first transcript, end last transcript). Plot a correlation between the number of genes and the lenght for each chromosome. You will need to perform a join.

 E) `exonsBy()` groups exons by gene or transcript; extract exons grouped by gene. (Challenging!) can you identify genes with exons on different chromosomes? Are there any of these genes on the standard chromosomes?
   
```{r}
    exByGn <- exonsBy(txdb, "gene")
    ##
    trans <- lengths(unique(seqnames(exByGn)))
    table( trans )
    seqnames( exByGn[ trans > 1 ] )
    ##
    std <- paste0("chr", c(1:22, "X", "Y", "M"))
    unames <- unique(seqnames(exByGn[ trans > 1 ]))
    transstd <- all(unames %in% std)
x <- unames[transstd]
```

 F) The previous exercise indicated that gene `"22947"` has exons on both chromosomes 4 and 10. Find out more about this gene using the [org.Hs.eg.db][] package and by searching for the gene symbol on the NCBI web site.

```{r}
    egid <- "22947"
    AnnotationDbi::select(
        org.Hs.eg.db, egid, c("SYMBOL", "GENENAME"), "ENTREZID"
)
```

 G) Using an entrezid you can directly view the the information on this gene from the NCBI website. Run the code below, what is the function of entregene `r egid`
    
```{r, eval = FALSE}
    url <- paste0("https://www.ncbi.nlm.nih.gov/gene/", egid)
    browseURL(url)
```
    
 H) Note that the `TxDb.*` packages also support `keytypes()`, `columns()`, and `select()` for mapping between exon, cds, transcript, and gene identifiers.
    
**Exercise 5**: `BSgenome.*` packages

 A) Load the [BSgenome.Hsapiens.UCSC.hg38][] package, containing the entire sequence of the hg38 build of _Homo sapiens_. Check out it's contents, and create a simple alias.
   
```{r}
    library(BSgenome.Hsapiens.UCSC.hg38)
    BSgenome.Hsapiens.UCSC.hg38
    hg38 <- BSgenome.Hsapiens.UCSC.hg38
```
   
 B) Genomic sequence can be retrieved by chromosome, e.g., `hg38[["chr1"]]`, or by genomic range, e.g., `getSeq(hg38, GRanges("chr1:1000000-2000000"))`. Retrieve the chunk ("chr1:1000000-2000000") of DNA and calculate GC content (`letterFrequency()`).
   
```{r}
    dna <- getSeq(hg38, GRanges("chr1:1000000-2000000"))
    letterFrequency(dna, "GC", as.prob=TRUE)
```

 C) Use the `org.*`, `TxDb.*`, and `BSgenome.*` packages to retrieve
   the BRCA1 exon DNA sequence.
   
**TIPS** 
 - You will need the `mapsIds()` function   
 - You will need the `exonsBy()` function   
```{r}
    brca1_egid <- mapIds(org.Hs.eg.db, "BRCA1", "ENTREZID", "SYMBOL")
    brca1_exons <- exonsBy(txdb, "gene")[[brca1_egid]]
    getSeq(hg38, brca1_exons)
```

**Exercise 6**

```{r, eval=FALSE}

###################################################
### code chunk number 38: search-nrsf
###################################################
library(BSgenome.Hsapiens.UCSC.hg19)
nrsfHits <- matchPattern("TCAGCACCATGGACAG", Hsapiens[["chr1"]])
length(nrsfHits)  # number of hits


###################################################
### code chunk number 39: track-nrsf
###################################################
nrsfTrack <- GenomicData(ranges(nrsfHits), strand="+", chrom="chr1",
                         genome = "hg19")


###################################################
### code chunk number 40: browserView-nrsf (eval = FALSE)
###################################################
session <- browseGenome(nrsfTrack, range = range(nrsfTrack[1]) * -10)


###################################################
### code chunk number 41: rmsk.e2f3 (eval = FALSE)
###################################################
library (rtracklayer)
mySession = browserSession("UCSC")
genome(mySession) <- "hg19"
e2f3.tss.grange <- GRanges("chr6", IRanges(20400587, 20403336))
tbl.rmsk <- getTable(
ucscTableQuery(mySession, track="rmsk", 
              range=e2f3.tss.grange, table="rmsk"))


###################################################
### code chunk number 43: trackAndTableNameDiscovery (eval = FALSE)
###################################################
mySession <- browserSession ()
genome(mySession) <- "hg19"
# 177 tracks in October 2012
track.names <- trackNames(ucscTableQuery(mySession))
##   # chose a few tracks at random from this set, and discover how 
##   # many tables they hold
tracks <- track.names [c (99, 81, 150, 96, 90)]
sapply(tracks, function(track) {
      length(tableNames(ucscTableQuery(mySession, track=track)))
      })


###################################################
### code chunk number 44: session-info
###################################################
  sessionInfo()


```




<font color="red">Internet access required for this exercise</font>

Here we use [rtracklayer][] to retrieve estrogen receptor binding
sites identified across cell lines in the ENCODE project. We focus on
binding sites in the vicinity of a particularly interesting region.

A) Define our region of interest by creating a `GRanges` instance with appropriate genomic coordinates. Our region corresponds to 10Mb up-and down-stream of a particular gene.
```{r}
roi <- GRanges("chr10", IRanges(92106877, 112106876, names="ENSG00000099194"))
```

B) Create a session for the UCSC genome browser
```{r rtracklayer-session, eval=FALSE}
library(rtracklayer) 
session <- browserSession()
``` 

C) Query the UCSC genome browser for ENCODE estrogen receptor ERalpha<sub>a</sub> transcription marks; identifying the appropriate track, table, and transcription factor requires biological knowledge and detective work.
```{r rtracklayer-marks, eval=FALSE}
trackName <- "wgEncodeRegTfbsClusteredV2"
tableName <- "wgEncodeRegTfbsClusteredV2"
trFactor <- "ERalpha_a"
ucscTable <- getTable(ucscTableQuery(session, track=trackName,
    range=roi, table=tableName, name=trFactor))
``` 

D) Visualize the location of the binding sites and their scores; annotate the mid-point of the region of interest.
```{r rtracklayer-plot, fig.height=3, eval=FALSE}

plot(score ~ chromStart, ucscTable, pch="+")
abline(v=start(roi) + (end(roi) - start(roi) + 1) / 2, col="blue")
``` 

# AnnotationHub

[AnnotationHub][] is a data base of large-scale whole-genome
resources, e.g., regulatory elements from the Roadmap Epigenomics
project, Ensembl GTF and FASTA files for model and other organisms,
and the NHLBI [grasp2db][] data base of GWAS results. There are many interesting ways in which these resources can be used. Examples include

- Easily access and import Roadmap Epigenomics files.
- 'liftOver' genomic range-based annotations from one coordinate
  system (e.g, hg38) to another (e.g., GRCh 38);
- Create TranscriptDb and BSgenome-style annotation resources 'on the
  fly' for a diverse set of organisms.
- Programmatically access the genomic coordiantes of clinically
  relevant variants cataloged in dbSNP.
  
Unfortunately, [AnnotationHub][] makes extensive use of internet
resources and so we will not pursue it in this course; see the
vignettes that come with the pacakge, for instance
[_AnnotationHub_ HOW-TOs][AH-howto].

[AH-howto]: http://bioconductor.org/packages/devel/bioc/vignettes/AnnotationHub/inst/doc/AnnotationHub-HOWTO.html

# Annotating variants

_Bioconductor_ provides facilities for reading VCF files. These work
very well with the annotation resources described above, so for
instance it is straight-forward to identify variants in coding or
other regions of interest.

To develop a sense of the capabilities available, work through the
[VariantAnnotation][] vignette 'Introduction to Variant Annotation',
and the [VariantFiltering][] vignette.

[AnnotationDbi]: http://bioconductor.org/packages/AnnotationDbi
[AnnotationHub]: http://bioconductor.org/packages/AnnotationHub
[BSgenome]: http://bioconductor.org/packages/release/BSgenome
[Bsgenome.Hsapiens.UCSC.hg38]: http://bioconductor.org/packages/Bsgenome.Hsapiens.UCSC.hg38
[grasp2db]: http://bioconductor.org/packages/release/grasp2db
[Gviz]: http://bioconductor.org/packages/release/Gviz
[Homo.sapiens]: http://bioconductor.org/packages/release/Homo.sapiens
[Rsamtools]: http://bioconductor.org/packages/Rsamtools
[TxDb.Hsapiens.UCSC.hg38.knownGene]: http://bioconductor.org/packages/TxDb.Hsapiens.UCSC.hg38.knownGene
[VariantAnnotation]: http://bioconductor.org/packages/VariantAnnotation
[VariantFiltering]: http://bioconductor.org/packages/VariantFiltering
[biomaRt]: http://bioconductor.org/packages/biomaRt
[org.Hs.eg.db]: http://bioconductor.org/packages/org.Hs.eg.db
[org.Sc.sgd.db]: http://bioconductor.org/packages/org.Sc.sgd.db
[rtracklayer]: http://bioconductor.org/packages/release/rtracklayer
[EnsDb.Hsapiens.v75]: http://bioconductor.org/packages/EnsDb.Hsapiens.v75

# End matter

## Session Info

```{r}
sessionInfo()
```

## Acknowledgements

Materials adapted from the CSAMA 2018 course

Research reported in this tutorial was supported by the National Human
Genome Research Institute and the National Cancer Institute of the
National Institutes of Health under award numbers U41HG004059 and
U24CA180996.

This project has received funding from the European Research Council
(ERC) under the European Union's Horizon 2020 research and innovation
programme (grant agreement number 633974)
