---
title: "GEOQuery"
author: "Marc A.T. Teunis"
date: "`r Sys.Date()`"
output:
  word_document:
  ioslides_presentation:
    highlight: pygments
    widescreen: yes
---
<style>

.footer {
    font-size: 16pt;
    color: DarkBlue;
    content: attr(label);
    position: absolute;
    margin-left: auto;
    margin-right: auto;
    left: 0;
    right: 0;
    text-align:center;
    bottom: 10px;
    line-height: 1.9;
    display: block;
}

</style>

<div class="footer" style="margin-top;font-size:120%;"> 
R & Bioconductor for Bioinformatics - May 2018 </div>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE,
                      message = FALSE,
                      error = FALSE)
```

```{r, root_1, include=FALSE}
## defines the root of the project for later use
require("rprojroot") || utils::install.packages("rprojroot")
library(rprojroot)
root <- find_root_file(criterion = is_rstudio_project)
```

## Introduction

from: https://warwick.ac.uk/fac/sci/moac/people/students/peter_cock/r/geo/
and Vignette of `{GEOquery}`
 
Learn:

 - How to download data directly from the GEO database from R
 - How to inspect the different data objects
 - How to create relevant data objects from the downloaded datasets
 - How to process datasets from transcriptome Array and next-gen sequencing experiments
 
## Packages
```{r}
library(msa)
library(tidyverse)
library(biomaRt)
library(annotate)
library(Biostrings)
library(seqinr)
library(DECIPHER)
library(GEOquery)
library(Biobase)
```

## Vignettes
```{r, eval=FALSE}
browseVignettes("GEOquery")
```

## GEO datasets

 - GEO Platform (GPL). These files describe a particular type of microarray. They are annotation files.

 - GEO Sample (GSM). Files that contain all the data from the use of a single chip. For each gene there will be multiple scores including the main one, held in the VALUE column.

 - GEO Series (GSE). Lists of GSM files that together form a single experiment.

 - GEO Dataset (GDS). These are curated files that hold a summarised combination of a GSE file and its GSM files. They contain normalised expression levels for each gene from each sample (i.e. just the VALUE field from the GSM file). 

## Downloading GDS dataasets
```{r}
#Download GDS file, put it in the current directory, and load it:
gds858 <- getGEO('GDS858', destdir=".")

#Or, open an existing GDS file (even if its compressed):
gds858_local <- getGEO(filename='GDS858.soft.gz')
```

## Inspecting the contents of the dataset
```{r}
meta_data <- Meta(gds858) %>% print()

```

## The sample table (names)
```{r}
colnames(Table(gds858))
```

## The actual expression matrix with probe_ids and expression values
```{r}
Table(gds858) %>% as_tibble 
```

## Terun 


## R code from vignette
```{r}
gds <- getGEO("GDS507")
gds <- getGEO(filename=system.file("extdata/GDS507.soft.gz",package="GEOquery"))




## ------------------------------------------------------------------------
# If you have network access, the more typical way to do this
# would be to use this:
gds <- getGEO("GSM11805")
gsm <- getGEO(filename=system.file("extdata/GSM11805.txt.gz",package="GEOquery"))
gsm
## ------------------------------------------------------------------------
# Look at gsm metadata:
head(Meta(gsm))
# Look at data associated with the GSM:
# but restrict to only first 5 rows, for brevity
Table(gsm)[1:5,]
# Look at Column descriptions:
Columns(gsm)

## ------------------------------------------------------------------------
Columns(gds)[1:3,]

## ------------------------------------------------------------------------
# Again, with good network access, one would do:
# gse <- getGEO("GSE781",GSEMatrix=FALSE)
gse <- getGEO(filename=system.file("extdata/GSE781_family.soft.gz",package="GEOquery"))
head(Meta(gse))
# names of all the GSM objects contained in the GSE
names(GSMList(gse))
# and get the first GSM object on the list
GSMList(gse)[[1]]
# and the names of the GPLs represented
names(GPLList(gse))

## ------------------------------------------------------------------------
# Note that GSEMatrix=TRUE is the default
gse2553 <- getGEO('GSE2553',GSEMatrix=TRUE)
show(gse2553)
show(pData(phenoData(gse2553[[1]]))[1:5,c(1,6,8)])

## ------------------------------------------------------------------------
eset <- GDS2eSet(gds,do.log2=TRUE)

## ------------------------------------------------------------------------
eset
pData(eset)[,1:3]

## ------------------------------------------------------------------------
#get the platform from the GDS metadata
Meta(gds)$platform
#So use this information in a call to getGEO
gpl <- getGEO(filename=system.file("extdata/GPL97.annot.gz",package="GEOquery"))

## ------------------------------------------------------------------------
MA <- GDS2MA(gds,GPL=gpl)
class(MA)

## ------------------------------------------------------------------------
gsmplatforms <- lapply(GSMList(gse),function(x) {Meta(x)$platform_id})
head(gsmplatforms)

## ------------------------------------------------------------------------
gsmlist = Filter(function(gsm) {Meta(gsm)$platform_id=='GPL96'},GSMList(gse))
length(gsmlist)

## ------------------------------------------------------------------------
Table(gsmlist[[1]])[1:5,]
# and get the column descriptions
Columns(gsmlist[[1]])[1:5,]

## ------------------------------------------------------------------------
# get the probeset ordering
probesets <- Table(GPLList(gse)[[1]])$ID
# make the data matrix from the VALUE columns from each GSM
# being careful to match the order of the probesets in the platform
# with those in the GSMs
data.matrix <- do.call('cbind',lapply(gsmlist,function(x) 
                                      {tab <- Table(x)
                                       mymatch <- match(probesets,tab$ID_REF)
                                       return(tab$VALUE[mymatch])
                                     }))
data.matrix <- apply(data.matrix,2,function(x) {as.numeric(as.character(x))})
data.matrix <- log2(data.matrix)
data.matrix[1:5,]

## ------------------------------------------------------------------------
require(Biobase)
# go through the necessary steps to make a compliant ExpressionSet
rownames(data.matrix) <- probesets
colnames(data.matrix) <- names(gsmlist)
pdata <- data.frame(samples=names(gsmlist))
rownames(pdata) <- names(gsmlist)
pheno <- as(pdata,"AnnotatedDataFrame")
eset2 <- new('ExpressionSet',exprs=data.matrix,phenoData=pheno)
eset2

## ------------------------------------------------------------------------
gpl97 <- getGEO('GPL97')
Meta(gpl97)$title
head(Meta(gpl97)$series_id)
length(Meta(gpl97)$series_id)
head(Meta(gpl97)$sample_id)
length(Meta(gpl97)$sample_id)

## ------------------------------------------------------------------------
gsmids <- Meta(gpl97)$sample_id
gsmlist <- sapply(gsmids[1:5],getGEO)
names(gsmlist)

## ----citation------------------------------------------------------------
citation("GEOquery")

## ----eval=FALSE----------------------------------------------------------
#  bug.report(package='GEOquery')

## ----echo=FALSE----------------------------------------------------------
sessionInfo()


```

# Exercise

## Literature
The article can be found at: http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006512
the dataset is available from:  GEO database (accession number GSE76073).


## Data
```{r}
gds_76073 <- getGEO("GSE76073", GSEMatrix = TRUE)
```

## Inspecting the data
Printing the gds object reveals two different `ExpressionSet` objects in a list
We can store them individually in two new variables.

```{r}
gds_76073

```



## RNAseq Platform



## Getting counts (experimental data)








