---
title: "Introduction to Bioconductor"
author: "Marc Teunis <marc.teunis@hu.nl>"
date: "`r Sys.Date()`"
output:
  ioslides_presentation: 
    highlight: pygments
    widescreen: yes
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE,
                      error = FALSE,
                      message = FALSE)

```

```{r, define_root, include=FALSE}
## defines the root of the project for later use
require("rprojroot") || utils::install.packages("rprojroot")
library(rprojroot)
root <- find_root_file(criterion = is_rstudio_project)
```

```{r pkgs, echo=FALSE}
suppressPackageStartupMessages({
    library(Biostrings)
    library(GenomicRanges)
    library(SummarizedExperiment)
    library(tidyverse)
})
```

## Credits
Part of this tutorial was reproduced and adpated from the CSAMA 2018 Course "Statistical analysis for Genome Scale Biology"
see https://github.com/bioconductor/CSAMA and http://www.huber.embl.de/csama2018/ for more details

Wilson G, Bryan J, Cranston K, Kitzes J, Nederbragt L, et al. (2017) Good enough practices in scientific computing. PLOS Computational Biology 13(6): e1005510.

    https://doi.org/10.1371/journal.pcbi.1005510
    http://bit.ly/good-enuff


## "Bioconductor"

Started 2002 as a platform for understanding analysis of microarray
data

More than 1,500 packages. Domains of expertise:

- Sequencing (RNASeq, ChIPSeq, single-cell, called variants, ...)
- Microarrays (methylation, expression, copy number, ...)
- flow cytometry
- proteomics
- ...

## Important themes

- Reproducible research
- Interoperability between packages & work kflows
- Usability

## Resources

- https://bioconductor.org
- https://bioconductor.org/packages -- software, annotation, experiment, workflow
- https://support.bioconductor.org

## High-throughput sequence work flow
```{r, echo=FALSE}
knitr::include_graphics(file.path(root,
                                  "images",
                                  "SequencingEcosystem.png"))
```

## DNA Sequences

[Biostrings][] themes

- _Valid_ data, e.g., alphabet
- 'Vector' interface: `length()`, `[`, ...
- Specialized operations, e.g,. `reverseComplement()`

## `DNAStringSet()`
```{r}
library(Biostrings)

dna <- DNAStringSet(c("AAACTG", "CCCTTCAAC", "TACGAA"))
dna
```

## Subsetting 
```{r}
length(dna)
dna[c(1, 3, 1)]
width(dna)
```

## `reverseComplement()`
```{r}
reverseComplement(dna)
```

## Getting Help!
```{r, eval=FALSE}
methods(class="DNAStringSet")
?"DNAStringSet"
browseVignettes(package="Biostrings")
```

## **EXERCISE**
How would you translate this sequence to a peptide?

```
ATGGGCGCATTTAATAGCGCGATACCTCTG
```

## Genomic Ranges

![](`r here::here("images", "GRanges.png")`)

[GenomicRanges][]

- Data (e.g., aligned reads, called peaks, copy number)
- Annotations (e.g., genes, exons, transcripts)
- Close relation to BED files (see `rtracklayer::import.bed()` and
  [HelloRanges][])
- Also vector interface -- `length()`, `[`, etc.

## Creating a `GenomicRanges`
```{r}
library(GenomicRanges)
gr <- GRanges(c("chr1:100-120", "chr1:115-130", "chr2:200-220"))
gr
```

## `Rle()` "Run-length encoding" 
```{r}
dna_sequence <- DNAString(
  "AAATATGCGCGTTACCCAGGGCTACGTCGCGCTAAAACCCCCGCGT"
  )
as(dna_sequence, "Rle")
```

## Intra-range operations

- e.g., `shift()`, `narrow()`, `flank()`

## Inter-range operations

- e.g., `reduce()`, `coverage()`, `gaps()`, `disjoin()`

## Between-range operations

- e.g., `countOverlaps()`, `findOverlaps()`, `summarizeOverlaps()`

## `shift()`
Shift sequence by +1 position
```{r}
shift(gr, 1)
```

## `countPverlaps()`
```{r}
query <- GRanges(c("chr1:4-6", "chr2:4-10"))
```

## Define subject
```{r}
subject <- GRanges(c("chr1:1-3", "chr2:4-5", "chr3:3-11"))
countOverlaps(subject, query)
```

## `findOverlaps()`
```{r}
findOverlaps(subject = subject, query = query)
intersect(subject, query)
```
see also: https://bedtools.readthedocs.io/en/latest/content/tools/intersect.html

## Visualize `GRanges`
```{r}
library(AneuFinder)
library(humarray)
plot.GRanges(subject)
## PACKAGE REMOVEd FROM CRAN
## instal old versions manualy from source
```
see: https://cran.r-project.org/src/contrib/Archive/humarray/

## Help!
```{r, eval=FALSE}
methods(class="GRanges")
methods(class="GRangesList")
?"GRanges"
?"GRangesList"
browseVignettes(package="GenomicRanges")
```

## Lists of Genomic Ranges
```{r}
 knitr::include_graphics(file.path(
   root,
   "images",
   "GRangesList.png"))
```
- e.g., exons-within-transcripts, alignments-within-reads

## Matching sequences
```{r}
seq <- DNAString("AATAGGCGCGATA")
matchPattern("AA", seq)
```

## Multiple patterns
```{r}
map(c("AA", "TA", "GC"), 
    matchPattern, 
    subject = seq)
```

## Multiple subjects
```{r}
vmatchPattern(pattern = "AA", 
              subject = DNAStringSet(
                c("AAAAATCCGC", 
                  "AATAGGCGCGATA")))
```

## Multiple subjects, multiple patterns
```{r}

patterns <- c("AA", "TA", "GC") 
subjects <- DNAStringSet(c("AAAAATCCGC", 
                    "AATAGGCGCGATA"))
```

## Provide names to better read the output
```{r}
names(subjects) <- c("peptide_1", "peptide_2") 
names(patterns) <- patterns 
```

## Create anonymous function
```{r}
anonymous_f <- function(x, y){
  vmatchPattern(pattern = x, 
                subject = y)
    
}
```

## Loop function over patterns and subjects
```{r}
map(patterns, anonymous_f, y = subjects)
```

## `SummarizedExperiment`

![](`r here::here("/images/SummarizedExperiment.png")`)

## Component parts

- three components -- underlying 'matrix', 'row' annotations (genomic
  features), 'column' annotations (sample descriptions)

```{r}
counts <- read.csv(here::here("data/airway_counts.csv"), row.names=1)
counts <- as.matrix(counts)
head(counts, 3)

colData <- read.csv(here::here("data/airway_colData.csv"), row.names=1)
colData[, 1:4]

rowRanges <- readRDS(here::here("data/airway_rowRanges.rds"))
updateObject(rowRanges)

```

## We could manipulate the information in the se object independently...
```{r, fig.asp=1}
cidx <- colData$dex == "trt"
plot(
    rowMeans(1 + counts[, cidx]) ~ rowMeans(1 + counts[, !cidx]),
    log="xy"
)
```

- very fragile, e.g., what if `colData` rows had been re-ordered?

Solution: Combine info in a single object -- [SummarizedExperiment][]

```{r, fig.asp=1}
library(SummarizedExperiment)

se <- SummarizedExperiment(counts, rowRanges = rowRanges, colData = colData)
cidx <- se$dex == "trt"
```

## `{tidyverse}` way
```{r}
library(tidyverse)
## ggplot

counts_df <- assay(se) %>%
  as_tibble

trt <- counts_df[, cidx] %>% 
  mutate(treatment = "trt",
         ensembl_id = rownames(assay(se))) %>%
  gather(1:4, 
         key = "sample_name",
         value = "counts") %>%
  mutate(counts = 1 + counts)
  
untrt <- counts_df[, !cidx] %>%
  mutate(treatment = "untrt",
         ensembl_id = rownames(assay(se))) %>%
  gather(1:4, 
         key = "sample_name",
         value = "counts") %>%
         mutate(counts = 1 + counts)
  
bind_df_summarized <- dplyr::bind_rows(trt, untrt) %>%
  group_by(treatment, ensembl_id) %>%
  summarise(mean_counts = mean(counts))

bind_df_summarized %>%
  spread(key = treatment, value = mean_counts) %>%
  ggplot(aes(x = log10(trt),
           y = log10(untrt))) +
  geom_point(alpha = 0.03)
       
```

## **EXERCISE**
Convert the above `tidyverse` way to a function that takes the `assay(se)` matrix and the coldata group information ("trt", "untrt" in the example) as input and generates the plot as output. How would you call this function?

Use this template

```
function_name <- function(assay_data, group_1, group_2){

  __function_body__

  plot <- ....

return(plot)
}
```



- Much more robust to 'bookkeeping' errors
- matrix-like interface: `dim()`, two-dimensional `[`, ...
- accessors: `assay()`, `rowData()` / `rowRanges()`, `colData()`, ...

## Getting Help!

```{r, eval=FALSE}
methods(class="SummarizedExperiment")
?"SummarizedExperiment"
browseVignettes(package="SummarizedExperiment")
```

## _Bioconductor_ packages for high-throughput genomic analysis

Web site, https://bioconductor.org

1562 'software' packages, https://bioconductor.org/packages

- Sequence analysis (RNASeq, ChIPSeq, called variants, copy number, 
  single cell)
- Microarrays (methylation, copy number, classical expression, ...)
- Annotation (more about annotations later this morning...)
- Flow cytometry
- Proteomics, image analysis, ...

## Discovery and use, e.g., [DESeq2][]

- https://bioconductor.org/packages
- Landing pages: title, description (abstract), installation instructions,
  badges
- Vignettes!
- [Workflows][], [course material][], [videos][], ...

## Support site, https://support.bioconductor.org

[Biostrings]: https://bioconductor.org/packages/Biostrings
[GenomicRanges]: https://bioconductor.org/packages/GenomicRanges
[rtracklayer]: https://bioconductor.org/packages/rtracklayer
[HelloRanges]: https://bioconductor.org/packages/HelloRanges
[SummarizedExperiment]: https://bioconductor.org/packages/SummarizedExperiment
[DESeq2]: https://bioconductor.org/packages/DESeq2
[Workflows]: https://bioconductor.org/packages/release/BiocViews.html#___Workflow
[course material]: https://bioconductor.org/help/course-materials/
[videos]: https://www.youtube.com/user/bioconductor


## Vignettes {.build}

 * Vignettes are long explanations and demos of a package
 * A vignette contains examples and a workflow that shows how the package can be used and which (research) questions can be adressed with the functions and datasets in the package
 * It usually is a good place to start for examples
 * It also shows the so-called dependencies of a package
 * It explains which other packages you would need and how the data should look to be able to work with the package the vignette belongs to

# End matter

## Session Info

```{r}
sessionInfo()
```

## Acknowledgements

